contracts/src/plugin/PluginUUPSUpgradeable.sol

```diff
@@ -63,6 +106,86 @@ abstract contract PluginUUPSUpgradeable is
         return _getImplementation();
     }
 
+    /// @notice Sets the target to a new target (`newTarget`).
+    /// @param _targetConfig The target Config containing the address and operation type.
+    function _setTargetConfig(TargetConfig calldata _targetConfig) internal virtual {
+        TargetConfig memory previousTargetConfig = currentTargetConfig;
+
+        currentTargetConfig = _targetConfig;
+
+        // safety check to avoid setting dao as `target` with `delegatecall` operation
+        // as this would not work and cause the plugin to be bricked.
+        if (
+            _targetConfig.target.supportsInterface(type(IDAO).interfaceId) &&
+            _targetConfig.operation == Operation.DelegateCall
+        ) {
+            revert InvalidTargetConfig(_targetConfig);
+        }
+
+        emit TargetSet(previousTargetConfig, _targetConfig);
+    }
+
+    /// @notice Forwards the actions to the currently set `target` for the execution.
+    /// @param _callId Identifier for this execution.
+    /// @param _actions actions that will be eventually called.
+    /// @param _allowFailureMap Bitmap-encoded number. TODO:
+    /// @return execResults address of the implementation contract.
+    /// @return failureMap address of the implementation contract.
+    function _execute(
+        bytes32 _callId,
+        IDAO.Action[] memory _actions,
+        uint256 _allowFailureMap
+    ) internal virtual returns (bytes[] memory execResults, uint256 failureMap) {
+        return
+            _execute(
+                currentTargetConfig.target,
+                _callId,
+                _actions,
+                _allowFailureMap,
+                currentTargetConfig.operation
+            );
+    }
+
+    /// @notice Forwards the actions to the `target` for the execution.
+    /// @param _target Forwards the actions to the specific target.
+    /// @param _callId Identifier for this execution.
+    /// @param _actions actions that will be eventually called.
+    /// @param _allowFailureMap Bitmap-encoded number. TODO:
+    /// @return execResults address of the implementation contract.
+    /// @return failureMap address of the implementation contract.
+    function _execute(
```

Comment:
We have quite some code duplication over all different Plugin* contracts.. probably worth refactoring it to have those methods only in one place..